File: src/main/java/ru/romanow/jpa/mapper/utils/ReferenceMapper.java
package ru.romanow.jpa.mapper.utils;

import lombok.SneakyThrows;
import org.mapstruct.ObjectFactory;
import org.mapstruct.TargetType;
import org.springframework.stereotype.Component;
import ru.romanow.jpa.model.IdentifiableModel;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

@Component
public class ReferenceMapper {

    @PersistenceContext
    private EntityManager entityManager;

    @ObjectFactory
    public <T> T resolve(IdentifiableModel model, @TargetType Class<T> type) {
        if (model.getId() != null) {
            final var entity = entityManager.getReference(type, model.getId());
            return entity != null ? entity : newInstance(type);
        } else {
            return newInstance(type);
        }
    }

    @SneakyThrows
    private <T> T newInstance(Class<T> type) {
        return type.getDeclaredConstructor().newInstance();
    }
}

File: src/main/java/ru/romanow/jpa/mapper/utils/ReferenceMapper.kt
package ru.romanow.jpa.mapper.utils

import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.mapstruct.ObjectFactory
import org.mapstruct.TargetType
import org.springframework.stereotype.Component
import ru.romanow.jpa.model.IdentifiableModel

@Component
class ReferenceMapper(@PersistenceContext private val entityManager: EntityManager) {

    @ObjectFactory
    fun <T> resolve(model: IdentifiableModel, @TargetType type: Class<T>): T {
        if (model.id != null) {
            val entity: T = entityManager.getReference(type, model.id)
            return entity ?: newInstance(type)
        } else {
            return newInstance(type)
        }
    }

    private fun <T> newInstance(type: Class<T>): T {
        return type.getDeclaredConstructor().newInstance()
    }
}
